### PR TITLE
DDCNL-3572 CySP COPE Enhanced Solution

### PR DESCRIPTION
**Pull request by gianick**

### CODE REVIEW

File: app/uk/gov/hmrc/statepension/controllers/ErrorHandling.scala [hunk 1]
```diff
@@ -17,48 +17,81 @@
...
+    func.recoverWith {
+      case WithStatusCode(NOT_FOUND, _) => Future.successful(NotFound(Json.toJson[ErrorResponse](ErrorNotFound)))
+      case _: NotFoundException => Future.successful(NotFound(Json.toJson[ErrorResponse](ErrorNotFound)))
+      case WithStatusCode(GATEWAY_TIMEOUT, e) => logger.error(s"$app Gateway Timeout: ${e.getMessage}", e); Future.successful(GatewayTimeout)
+      case WithStatusCode(BAD_GATEWAY, e) => logger.error(s"$app Bad Gateway: ${e.getMessage}", e); Future.successful(BadGateway)
+      case WithStatusCode(BAD_REQUEST, _) =>
+        Future.successful(BadRequest(Json.toJson(ErrorGenericBadRequest("Upstream Bad Request. Is this customer below State Pension Age?"))))
+      case WithStatusCode(UNPROCESSABLE_ENTITY, e) if e.message.contains("NO_OPEN_COPE_WORK_ITEM") => defineCopeResponse(nino)
+      case WithStatusCode(UNPROCESSABLE_ENTITY, e) if  e.message.contains("CLOSED_COPE_WORK_ITEM") =>
+        copeRepository.delete(HashedNino(nino))
```
Comment (kurtlogan) [inline app/uk/gov/hmrc/statepension/controllers/ErrorHandling.scala:59]: Need to wait for this future to complete. Use a map to do this

File: app/uk/gov/hmrc/statepension/repositories/CopeRepository.scala [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import uk.gov.hmrc.mongo.MongoComponent
+import uk.gov.hmrc.mongo.play.json.{Codecs, PlayMongoRepository}
+import uk.gov.hmrc.statepension.config.AppConfig
+import uk.gov.hmrc.statepension.controllers.HashedNino
+import uk.gov.hmrc.statepension.models.CopeRecord
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class CopeRepository @Inject()(mongo: MongoComponent)(implicit ec: ExecutionContext, appConfig: AppConfig)
+  extends PlayMongoRepository[CopeRecord](
```
Comment (kurtlogan) [inline app/uk/gov/hmrc/statepension/repositories/CopeRepository.scala:37]: ttl for record?

File: conf/application.conf [hunk 3]
```diff
@@ -198,6 +203,7 @@ microservice {
 # ~~~~~
 cope {
     dwp.originatorId = "DA_PFDWP"
-    feature.enabled = true
-    returnToServiceDays = 91
+    returnToServiceWeeks = 4
 }
+
+ninoHashingKey = "gn:a$V96?cFHmjjcSz31tU,Xe;hOLIYeknIPIbj6Y2GAyht3#Cs+ir&hd{(50`vGAd/k8OIIV:nbM4,c&z9rX+mmK<?qK`sbmKB*Y#Ip.N'$@sX~W[6{KIXA|k`"
```
Comment (kurtlogan) [inline conf/application.conf:209]: Raise a PR for production so this is different in there

**Review by kurtlogan — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/statepension/repositories/CopeRepository.scala (hunk 2 above)
Comment (gianick) [reply app/uk/gov/hmrc/statepension/repositories/CopeRepository.scala:38]: Added a TTL index and tested it. Deleted DB documents after the expiry date.

File: conf/application.conf (hunk 3 above)
Comment (Tateni93) [reply conf/application.conf:210]: https://github.com/hmrc/app-config-production/pull/10990 Done

File: app/uk/gov/hmrc/statepension/controllers/ErrorHandling.scala (hunk 1 above)
Comment (Tateni93) [reply app/uk/gov/hmrc/statepension/controllers/ErrorHandling.scala]: Done

**Review by kurtlogan — APPROVED** [review]

**Pull request merged**